File: leetcode-cn/783.二叉搜索树节点最小距离.py
#
# @lc app=leetcode.cn id=783 lang=python3
#
# [783] 二叉搜索树节点最小距离
#


# @lc code=start
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def minDiffInBST(self, root: TreeNode) -> int:
        pre = 9999
        ans = 9999

        def help(root):
            if root is None:
                return

            nonlocal ans, pre

            help(root.left)

            val = root.val
            ans = min(ans, abs(val - pre))
            pre = val

            help(root.right)

        help(root)
        return ans


# @lc code=end

File: leetcode-cn/C++/783.二叉搜索树节点最小距离.cpp
/*
 * @lc app=leetcode.cn id=783 lang=cpp
 *
 * [783] 二叉搜索树节点最小距离
 */

// @lc code=start
#include <algorithm>
#include <iostream>
using namespace std;

struct TreeNode {
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x)
        : val(x)
        , left(NULL)
        , right(NULL)
    {
    }
};

class Solution {
public:
    int pre = -1;
    int ans = INT32_MAX;

    void dfs(TreeNode* root)
    {
        if (root == NULL) {
            return;
        }

        dfs(root->left);

        if (pre != -1) {
            ans = min(ans, root->val - pre);
        }
        pre = root->val;

        dfs(root->right);
    }

    int minDiffInBST(TreeNode* root)
    {
        dfs(root);
        return ans;
    }
};
// @lc code=end
